Porting to ASP.NET Core (#21042)
dotnet/docs: docs/architecture/porting-existing-aspnet-apps/incremental-migration-strategies.md

---
title: Strategies for migrating incrementally
description: What strategies can teams adopt that will allow them to migrate large applications from ASP.NET MVC to .NET Core in an incremental fashion?
author: ardalis
ms.date: 11/13/2020
---

# Strategies for migrating incrementally

The biggest challenge with migrating any large application is figuring out how to break the very large process up into smaller pieces. There are several strategies that can be applied for this purpose, including breaking up the app into horizontal layers such as UI, data access, business logic, or breaking up the app into separate, smaller apps. Another strategy that can help in this regard is to upgrade some or all of the application to different framework versions on the way to a recent .NET Core release.

Consider the challenge of migrating a large ASP.NET 4.5 application. One approach would be to migrate directly from .NET 4.5 to .NET Core 3.1. However, this approach would need to account for every breaking change between these two frameworks and versions, which are substantial.

> One recent addition to the .NET ecosystem that helps with interoperability between different .NET frameworks is [.NET Standard](https://dotnet.microsoft.com/platform/dotnet-standard). The 2.0 .NET Standard release is notable because covers the vast majority of base class library functionality used by most .NET and .NET Core applications. Unfortunately, the earliest version of .NET with support for .NET Standard 2.0 is .NET 4.6.1.

Reviewer: Could we write a one-liner from the dotnet-standard doc that expplain what it is?

> NET Standard allows libraries to build against the agreed on set of common APIs, ensuring they can be used in any .NET application.

Reviewer: Could we list it as `.NET Standard 2.0`? I've not encountered the version number appear first elsewhere.

~~The 2.0 .NET Standard release~~ .NET Standard 2.0 is notable because it covers ...

> One recent addition to the .NET ecosystem that helps with interoperability between different .NET frameworks is [.NET Standard](https://dotnet.microsoft.com/platform/dotnet-standard). .NET Standard allows libraries to build against the agreed on set of common APIs, ensuring they can be used in any .NET application. .NET Standard 2.0 is notable because covers the vast majority of base class library functionality used by most .NET and .NET Core applications. Unfortunately, the earliest version of .NET with support for .NET Standard 2.0 is .NET 4.6.1.

Reviewer 2: ```suggestion
One recent addition to the .NET ecosystem that helps with interoperability between different .NET frameworks is [.NET Standard](https://dotnet.microsoft.com/platform/dotnet-standard). .NET Standard allows libraries to build against the agreed on set of common APIs, ensuring they can be used in any .NET application. .NET Standard 2.0 is notable because covers the vast majority of base class library functionality used by most .NET and .NET Core applications. Unfortunately, the earliest version of .NET with support for .NET Standard 2.0 is .NET Framework 4.6.1.
```

> Thus, one approach to incrementally upgrade a .NET 4.5 system would be to first update its class libraries to .NET 4.6.1 and then to modify these libraries to be .NET Standard class libraries (using multi-targeting and conditional compilation, if required). Such libraries can be consumed by ASP.NET Core 2.1 apps, so the next step would be to migrate some or all of the web functionality of the app to ASP.NET Core 2.1 (as described in the [previous chapter](choose-net-core-version.md)).

Reviewer 3: Might be good to docs on multi-targeting for those who want to learn more.

> Thus, one approach to incrementally upgrade a .NET 4.5 system would be to first update its class libraries to .NET 4.6.1 and then to modify these libraries to be .NET Standard class libraries (using multi-targeting and conditional compilation, if required). Such libraries can be consumed by ASP.NET Core 2.1 apps, so the next step would be to migrate some or all of the web functionality of the app to ASP.NET Core 2.1 (as described in the [previous chapter](choose-net-core-version.md)).

Reviewer 3: If libraries target NetStandard 2.0, the web app consuming them can move straight to ASP.NET Core 3.1. I don't think there's much value in using 2.1 as a stepping stone unless the customer has to stay on .NET Framework for a while (which isn't the case if their libraries target netstandard).

> Thus, one approach to incrementally upgrade a .NET 4.5 system would be to first update its class libraries to .NET 4.6.1 and then to modify these libraries to be .NET Standard class libraries (using multi-targeting and conditional compilation, if required). This step can be helpful in scenarios where app dependencies require .NET Framework and cannot easily be ported directly to use .NET Standard and .NET Core. Since .NET Framework libraries can be consumed by ASP.NET Core 2.1 apps, the next step would be to migrate some or all of the web functionality of the app to ASP.NET Core 2.1 (as described in the [previous chapter](choose-net-core-version.md)).

Reviewer 2: ```suggestion
Thus, one approach to incrementally upgrade a .NET Framework 4.5 system would be to first update its class libraries to .NET 4.6.1 and then to modify these libraries to be .NET Standard class libraries (using multi-targeting and conditional compilation, if required). This step can be helpful in scenarios where app dependencies require .NET Framework and cannot easily be ported directly to use .NET Standard and .NET Core. Since .NET Framework libraries can be consumed by ASP.NET Core 2.1 apps, the next step would be to migrate some or all of the web functionality of the app to ASP.NET Core 2.1 (as described in the [previous chapter](choose-net-core-version.md)).
```

> Once the app is running on ASP.NET Core 2.1, migrating it to ASP.NET Core 3.1 in isolation is relatively straightforward. The most likely challenge that will occur during this step will be updating any incompatible dependencies to support .NET Core and possibly higher versions of .NET Standard.

Reviewer 3: I think it would be helpful to clarify that this is specifically for cases where there are problematic dependencies that need .NET Framework. If dependencies can target .NET Standard or Core I'd skip 2.1 altogether.

> By the time the application is running on .NET Core 3.1, migrating to the current .NET 5.0 release should be relatively painless. The process primarily involves updating the target framework of your project files and their associated NuGet package dependencies. While there are a number of [breaking changes to consider](/dotnet/core/compatibility/3.1-5.0), most applications shouldn't require substantial modification to move from .NET Core 3.1 to .NET 5.0.

Reviewer 3: This makes sense. Might be worth a short mention that not all customers will want to move to 5 (you can link to the 'choose the right .NET Core version' page).

> By the time the application is running on .NET Core 3.1, migrating to the current .NET 5 release should be relatively painless. The process primarily involves updating the target framework of your project files and their associated NuGet package dependencies. While there are a number of [breaking changes to consider](https://docs.microsoft.com/dotnet/core/compatibility/3.1-5.0), most applications should not require substantial modification to move from .NET Core 3.1 to .NET 5.

Reviewer 2: ```suggestion
By the time the application is running on .NET Core 3.1, migrating to the current .NET 5.0 release should be relatively painless. The process primarily involves updating the target framework of your project files and their associated NuGet package dependencies. While there are a number of [breaking changes to consider](/dotnet/core/compatibility/3.1-5.0), most applications shouldn't require substantial modification to move from .NET Core 3.1 to .NET 5.0.
```
